nemo/core/classes/modelPT.py

```diff
@@ -92,6 +92,8 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         if isinstance(self._trainer, Trainer):
             if self._trainer.num_gpus and self._trainer.num_nodes:
                 app_state.world_size = self._trainer.num_gpus * self._trainer.num_nodes
+            if torch.cuda.current_device() is not None:
```

Comment:
During restoration, trainer is not attached immediately, so the above if case will never be called. Maybe put it inside `ModelPT.set_trainer()` and call that here (first check if passed value is None or not)?

Reply:
@tkornuta-nvidia mentioned that we might be able to get num_gpus and num_nodes from hydra_runner. That info is being specified in the config, it would be great if hydra_runner could put it in AppState. What do you think?

Reply:
Right now, the decorator only injects hydra args. In the future there might be a case for making it do additional stuff, but for now I think it's best to deal with it at ModelPT or NLPModel level. Reason being is that in a non hydra script, these magic values will not be updated and cause issues to users - say in a notebook or colab environment which doesn't use hydra. Where as a ModelPT or NLPModel will still work without issue

Reply:
Can we take num_nodes and num_gpus directly from trainer instance instead of Hydra? That is the real "source of truth" since trainer is actually doing instance and gpu management

Reply:
@titu1994 "During restoration, trainer is not attached immediately, so the above if case will never be called."
Exactly. That is why I generally think relying on trainer instance makes no sense - as it is available only in some cases. 

@okuchaiev Regarding "source of truth": trainer or trainer configuration? Please note the arguments you listed (num_nodes and num_gpus) come clearly from trainer configuration. What else do we need? Number of epochs? we can retrieve them also from trainer configuration...

Moreover, there are some "important" variables that come from other sections of configuration (e.g. batch_size, skip_last from dataloader). For me this is yet another argument for having Application State that gathers them at the very start of the application. 

@ericharper Summarizing: according to my best knowledge, hydra runner seems like a perfect fit for me. The question is: what am I missing?

Reply:
@titu1994  "Reason being is that in a non hydra script, these magic values will not be updated and cause issues to users - say in a notebook or colab environment which doesn't use hydra"

I really do not think that it is a showstopper. We still are using PTL trainer and still have its configuration. Not seeing any real advantage of having Model.set_trainer() over having e.g. AppState.set_config(cfg)...

Reply:
@tkornuta-nvidia The issue here is not with trainer not being available immediately, but the fact that we want to support restoration - the correct way to require trainer in such a case is set_trainer(). Model restoration in DDP is quite a special case - do we want to restore in one or more GPUs, what if env doesn't have multiple GPUs (or just CPUs for that matter). These all need to be addressed in the future or in this PR itself. 

Trainer is required by multiple components throughout the lifetime of ModelPT - if the user wants to use everything by hand they are free to do so, and need not set trainer.

Regarding your comment about trainer or trainer configuration, we need trainer - the actual object which interprets those commands and sets the correct values inside itself. Slurm ddp is a variable set by PTL object which we will possibly never set inside a config/env variable (nor should we try to as it breaks PTL slurm).

Gpu resolution, node resolution all depend on object not config. So yes, we absolutely need the ptl trainer object throughout the lifecycle of ModelPT if it needs to interact anyway with the PTL ecosystem. 

Why is batch size a part of appstate. By its own definition, app state is monitoring global variables. Batch size is local to each data loader, or even multiple data loaders. ASR val and test data loaders reduce batch size cause they don't filter long audio clips. Same for the other local variables you mention which are all data loader contexts not global contexts. 

Hydra runner is not meant to update appstate. It's only rightful task by it's nomenclature is to manage hydra for DDP environments. 

This is because hydra runner runs at the pre ModelPT and trainer phase of execution - it simply doesn't have the scope to perform such bookkeeping for appstate.

And yes it is a show stopper - the most simplistic way (and what is shown by all of our tutorials) is to use notebooks to inspect and evaluate models or do some iterative development before shuttling it off to a hydra script. 

We use hydra, does not mean we force the user to use hydra in every part of their project too. 

I ask here, why bother with Appstate if trainer already gives you this information. If it's simply a singleton to hold and preserve information from trainer, that's fine. But trainer is true source of truth for current running environment, whereas appstate is just a replica of some snapshot of Trainer.

Reply:
I was maintaining AppState because modules don't have access to the Trainer object (only models do as far as I know).

Reply:
Good point, yes in that case appstate needs to be maintained. I am concerned as to how model restoration will work in such a case - the appstate will either be partially filled or be in the default state upon restoration, therefore module level objects will not be able to be instantiated until appstate is updated, which comes after set trainer

Reply:
We can alleviate this by taking a trainer during restoration optionally but it needs further discussion as to possible alternatives

Reply:
I will make sure model restoration works with model parallel. This PR is adding more basic functionality.

Reply:
Yup sounds good. Let's get @ekmb to sign off and merge.

Reply:
> Good point, yes in that case appstate needs to be maintained. I am concerned as to how model restoration will work in such a case - the appstate will either be partially filled or be in the default state upon restoration, therefore module level objects will not be able to be instantiated until appstate is updated, which comes after set trainer

There are several use-cases for model restoration, I agree this needs deeper thinking/discussion.